sha2/build.rs

```diff
@@ -1,14 +1,34 @@
+use std::collections::HashSet;
+
 fn main() {
     use std::env;
 
     let target_arch = env::var("CARGO_CFG_TARGET_ARCH").unwrap_or_default();
     let target_vendor = env::var("CARGO_CFG_TARGET_VENDOR").unwrap_or_default();
+    let features: HashSet<String> = env::var("CARGO_CFG_TARGET_FEATURE")
+        .unwrap_or_default()
+        .split(',')
+        .map(|x| x.to_string())
+        .collect();
 
     let mut build256 = cc::Build::new();
     let (sha256_path, sha512_path) = if target_arch == "x86" {
         ("src/sha256_x86.S", "src/sha512_x86.S")
     } else if target_arch == "x86_64" {
-        ("src/sha256_x64.S", "src/sha512_x64.S")
+        let sha512 = if features.contains("avx2") {
+            "src/sha512_x64_avx2.S"
+        } else {
+            "src/sha512_x64.S"
+        };
```

Comment:
I think we should link in both of these backends, and use CPU feature detection in order to select which one is best, rather than using compile-time gating.

We use runtime CPU feature detection in the `sha2` crate itself, and users may not necessarily compile with the `avx2` target feature enabled.

Reply:
Done